lib/improver/cli/__init__.py

```diff
@@ -0,0 +1,441 @@
+
+    legacy_main = getattr(sys.modules[obj.__module__], 'main', None)
+
+    if legacy_main is None or use_clize and obj.__annotations__:
+        # legacy_main is None here for any function defined in this module
+        # before (and including) the main function
+        return Clize.keep(obj, **kwargs)
+
+    def _wrapper(prog: parameters.pass_name, *args):
+        sys.argv[0] = prog.split()[-1]
+        legacy_main(args)
+
+    description = obj.__doc__.split('\n')[0].strip()
+    _obj = Clize.as_is(_wrapper, description=description)
+    obj.cli = _obj.cli
+
+    return obj
+
+
+def clizefy(func=None, helper_class=DocutilizeClizeHelp, **kwargs):
+    """Decorator for creating CLI objects.
+
+    The legacy interface expects `<cli_name>.main` routine and the new one
+    expects <cli_name.process routine that is type annotated.
+    If both interfaces are available, then one is picked based on the
+    IMPROVER_USE_CLIZE environment variable, default is the legacy one.
+    The environment setting has to be done before `import improver.cli`.
+
+    Args:
+        func (obj):
+            The function
+        helper_class:
+        **kwargs:
+
+    Returns:
+
+    """
+    from functools import partial
+    if func is None:
+        return partial(clizefy, helper_class=helper_class, **kwargs)
+    func = _clizefy(func, helper_class=helper_class, **kwargs)
+    return func
+
+
+# help command
+
+
+@clizefy(help_names=())
+def improver_help(prog_name: parameters.pass_name,
+                  command=None, *, usage=False):
+    """Show command help."""
+    prog_name = prog_name.split()[0]
+    args = filter(None, [command, '--help', usage and '--usage'])
+    result = execute_command(SUBCOMMANDS_DISPATCHER, prog_name, *args)
+    if not command and usage:
+        result = '\n'.join(line for line in result.splitlines()
+                           if not line.endswith('--help [--usage]'))
+    return result
+
+
+def _cli_items():
+    """Dynamically discover CLIs."""
+    import importlib
+    import pkgutil
+    from improver.cli import __path__ as improver_cli_pkg_path
+    yield ('help', improver_help)
+    for minfo in pkgutil.iter_modules(improver_cli_pkg_path):
+        mod_name = minfo.name
+        if mod_name != '__main__':
+            mcli = importlib.import_module('improver.cli.' + mod_name)
+            yield (mod_name, clizefy(mcli.process))
+
+
+SUBCOMMANDS_TABLE = OrderedDict(sorted(_cli_items()))
+
+
+# main CLI object with subcommands
+
+
+SUBCOMMANDS_DISPATCHER = clizefy(
+    SUBCOMMANDS_TABLE,
+    description="""IMPROVER NWP post-processing toolbox""",
+    footnotes="""See also improver --help for more information.""")
+
+
+# IMPROVER top level main
+
+
+def unbracket(args):
+    """Convert input list with bracketed items into nested lists.
+
+    unbracket(['foo', '[', 'bar', 'a', 'b', ']',
+                            '[', 'baz', 'c', ']', '-o', 'z'])
+    ['foo', ['bar', 'a', 'b'], ['baz', 'c'], '-o', 'z']
```

Comment:
@tjtg @arh89, should this adhere to doctest format as was in the original code?

Reply:
I think so, yeah (but at the moment, we don't actually run doctests)

Reply:
I have beeen made into a unit test.

Reply:
That's fine, but why not follow an established standard of formatting code examples if there is one?

Reply:
I have changed it back but added the example of how the cli breaks up to arguments because although i know that's how it works, I find it much easier to visualize.